src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,308 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Driver.Support;
+
+namespace MongoDB.Driver
+{
+    internal static class TransactionExecutor
+    {
+        // constants
+        private const string TransientTransactionErrorLabel = "TransientTransactionError";
+        private const string UnknownTransactionCommitResultLabel = "UnknownTransactionCommitResult";
+        private static readonly TimeSpan __withTransactionTimeout = TimeSpan.FromSeconds(120);
+
+        internal static TResult ExecuteWithRetries<TResult>(
+            IClientSessionHandle clientSession, 
+            Func<IClientSessionHandle, CancellationToken, TResult> callback, 
+            DateTime startTime,
+            IDateTimeProvider dateTimeProvider,
+            TransactionOptions transactionOptions, 
+            CancellationToken cancellationToken)
+        {
+            TResult result;
+
+            while (true)
+            {
+                clientSession.StartTransaction(transactionOptions);
+
+                var transactionStepResult = WithTransactionCallbackProcessing(clientSession, callback, startTime, dateTimeProvider, cancellationToken);
+                if (transactionStepResult.TransactionShouldBeRestarted)
+                {
+                    continue;
+                }
+
+                result = transactionStepResult.Result;
+
+                if (!clientSession.IsInTransaction)
+                {
+                    return result; // Assume callback intentionally ended the transaction.
+                }
+
+                var transactionHasBeenConfirmed = WithTransactionCommitProcessingWithRetry(clientSession, startTime, dateTimeProvider, cancellationToken);
+                if (transactionHasBeenConfirmed)
```

Comment:
Would it be more accurate to name this variable "transactionHasBeenCommitted"?

Reply:
Done